File: Rectangle.Windows/src/Rectangle.Windows/Services/TileAllManager.cs
using System;
using System.Collections.Generic;
using Windows.Win32.Foundation;
using Rectangle.Windows.Core;

namespace Rectangle.Windows.Services;

/// <summary>
/// 平铺所有窗口管理器
/// 将所有可见窗口平铺排列在屏幕上
/// </summary>
public class TileAllManager
{
    private readonly Win32WindowService _win32;
    private readonly ConfigService? _configService;

    public TileAllManager(Win32WindowService win32, ConfigService? configService = null)
    {
        _win32 = win32;
        _configService = configService;
    }

    /// <summary>
    /// 平铺所有窗口
    /// </summary>
    /// <param name="workArea">工作区域</param>
    public void TileAll(WorkArea workArea)
    {
        var windows = WindowEnumerator.EnumerateVisibleWindows();
        if (windows.Count == 0) return;

        // 计算最佳网格布局
        var (cols, rows) = CalculateGridLayout(windows.Count, workArea);

        // 计算每个单元格的大小
        var cellWidth = workArea.Width / cols;
        var cellHeight = workArea.Height / rows;

        // 应用布局
        int index = 0;
        foreach (var hwnd in windows)
        {
            if (index >= cols * rows) break;

            int col = index % cols;
            int row = index / cols;

            int x = workArea.Left + col * cellWidth;
            int y = workArea.Top + row * cellHeight;

            _win32.SetWindowRect(hwnd, x, y, cellWidth, cellHeight);
            index++;
        }

        Logger.Info("TileAllManager", $"平铺 {windows.Count} 个窗口，布局: {cols}x{rows}");
    }

    /// <summary>
    /// 平铺当前应用的所有窗口
    /// </summary>
    /// <param name="workArea">工作区域</param>
    /// <param name="processName">进程名称</param>
    public void TileActiveApp(WorkArea workArea, string processName)
    {
        var windows = WindowEnumerator.EnumerateWindowsByProcess(processName);
        if (windows.Count == 0) return;

        // 计算最佳网格布局
        var (cols, rows) = CalculateGridLayout(windows.Count, workArea);

        // 计算每个单元格的大小
        var cellWidth = workArea.Width / cols;
        var cellHeight = workArea.Height / rows;

        // 应用布局
        int index = 0;
        foreach (var hwnd in windows)
        {
            if (index >= cols * rows) break;

            int col = index % cols;
            int row = index / cols;

            int x = workArea.Left + col * cellWidth;
            int y = workArea.Top + row * cellHeight;

            _win32.SetWindowRect(hwnd, x, y, cellWidth, cellHeight);
            index++;
        }

        Logger.Info("TileAllManager", $"平铺应用 {processName} 的 {windows.Count} 个窗口，布局: {cols}x{rows}");
    }

    /// <summary>
    /// 计算最佳网格布局
    /// </summary>
    private (int cols, int rows) CalculateGridLayout(int windowCount, WorkArea workArea)
    {
        if (windowCount <= 1) return (1, 1);
        if (windowCount == 2) return (2, 1);
        if (windowCount == 3) return (3, 1);
        if (windowCount == 4) return (2, 2);
        if (windowCount <= 6) return (3, 2);
        if (windowCount <= 9) return (3, 3);
        if (windowCount <= 12) return (4, 3);
        if (windowCount <= 16) return (4, 4);

        // 更多窗口时按比例计算
        var aspectRatio = (double)workArea.Width / workArea.Height;
        var total = (double)windowCount;
        var cols = (int)Math.Ceiling(Math.Sqrt(total * aspectRatio));
        var rows = (int)Math.Ceiling(total / cols);

        return (cols, rows);
    }
}

File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Rust/src/services/tile_all_manager.rs
use crate::core::rect::WorkArea;
use crate::services::win32::Win32WindowService;
use crate::services::window_enumerator::WindowEnumerator;
use crate::services::config::ConfigService;
use crate::services::logger::Logger;

/// 平铺所有窗口管理器
pub struct TileAllManager {
    win32: Win32WindowService,
    config_service: Option<ConfigService>,
}

impl TileAllManager {
    pub fn new(win32: Win32WindowService, config_service: Option<ConfigService>) -> Self {
        Self { win32, config_service }
    }

    /// 平铺所有窗口
    pub fn tile_all(&self, work_area: &WorkArea) {
        let windows = WindowEnumerator::enumerate_visible_windows(true, true);
        if windows.is_empty() {
            return;
        }

        self.tile_windows(work_area, &windows);

        Logger::info("TileAllManager", &format!("平铺 {} 个窗口", windows.len()));
    }

    /// 平铺当前应用的所有窗口
    pub fn tile_active_app(&self, work_area: &WorkArea, process_name: &str) {
        let windows = WindowEnumerator::enumerate_windows_by_process(process_name, true);
        if windows.is_empty() {
            return;
        }

        self.tile_windows(work_area, &windows);

        Logger::info("TileAllManager", &format!("平铺应用 {} 的 {} 个窗口", process_name, windows.len()));
    }

    /// 平铺窗口（内部实现）
    fn tile_windows(&self, work_area: &WorkArea, windows: &[isize]) {
        let window_count = windows.len();
        let (cols, rows) = self.calculate_grid_layout(window_count, work_area);

        let cell_width = work_area.width() / cols;
        let cell_height = work_area.height() / rows;

        for (index, hwnd) in windows.iter().enumerate() {
            if index >= (cols * rows) as usize {
                break;
            }

            let col = (index as i32) % cols;
            let row = (index as i32) / cols;

            let x = work_area.left + col * cell_width;
            let y = work_area.top + row * cell_height;

            self.win32.set_window_rect(*hwnd, x, y, cell_width, cell_height);
        }
    }

    /// 计算最佳网格布局
    fn calculate_grid_layout(&self, window_count: usize, work_area: &WorkArea) -> (i32, i32) {
        let count = window_count as i32;

        match window_count {
            0 => (1, 1),
            1 => (1, 1),
            2 => (2, 1),
            3 => (3, 1),
            4 => (2, 2),
            5 | 6 => (3, 2),
            7 | 8 | 9 => (3, 3),
            10 | 11 | 12 => (4, 3),
            13 | 14 | 15 | 16 => (4, 4),
            _ => {
                // 更多窗口时按比例计算
                let aspect_ratio = work_area.width() as f64 / work_area.height() as f64;
                let total = count as f64;
                let cols = (total.sqrt() * aspect_ratio.sqrt()).ceil() as i32;
                let rows = (total / cols as f64).ceil() as i32;
                (cols.max(1), rows.max(1))
            }
        }
    }
}

impl Default for TileAllManager {
    fn default() -> Self {
        Self::new(Win32WindowService::new(), None)
    }
}
